### DISCUSSION TITLE
Roadmap, future directions and organization

### DISCUSSION THREAD

**Discussion by JeanKossaifi** [tensorly/tensorly#290] [Ideas]
Hi all!

As the project grows it might be handy to have a common platform to easily exchange and chat. Gitter isn't the best for that so I'm thinking of creating a Slack, if everybody is happy with that.

I also created https://github.com/tensorly/lab for quickly trying new things, for instance quickly iterating on the API of the object oriented interface @MarieRoald @yngvem or new optimization routines @cohenjer @caglayantuna -- this has the advantage of being pip-installable for quick experimentation on unstable functionalities, without messing up with the main package.

To get started on a roadmap, I think these are outstanding issues:

## New features
* [ ] More decompositions:
  -  BTD @juliagusak 
  -  Tensor-Ring @merajhashemi ✅  PR #229 
  -  CMTF @IsabellLehmann and @aarmey PR #293 
* [ ] Generalize tensor regression to tensor outputs (#51 )
* [ ] Add support for tensor classification (e.g. #273 ) 
* [ ] (hand-optimized) gradient of some functions (e.g. CP reconstruction error) @aarmey and @JeanKossaifi 
* [ ] Serialization of models @yngvem PR https://github.com/tensorly/lab/pull/4 
* [ ] Visualization submodule - rather than just examples, we could have some general visualization tools

## CI (Continuous Integration)
* [ ] Refactor the tests so they are more reliable, some tend to occasionally fail (e.g. #287 #288) - Tracked in #310 
* [ ] Update Conda build and have it automatically done on release in Github worflows

## Enhancements of existing functions
* [ ] Better support for missing values (e.g. #198, @aarmey )

## API Improvement 
* [ ] More consistent tensor product support ( #250 ) - PR #309 
* [ ] Better SVD interface (#251, #233, #227) 
* [ ] Refine the API:
  - Backend VS main functions (we may want to allow backends to define some higher level functions, e.g. #249)
  -  Factorized tensors: OO interface, general support for batching (#199) 
  - Tensor decomposition objects, better support/integration with transform and pipelines (e.g. #172 ). I am iterating in tensorly-torch with [a more feature rich version](http://tensorly.org/torch/dev/modules/generated/tltorch.FactorizedTensor.html#tltorch.FactorizedTensor)
* [ ] Sparse tensors: add support for more backends, should the sparse support be in a separate package (e.g. tensorly-sparse) or moved to a main submodule (e.g. tensorly.sparse)? @asmeurer 
* [ ] Formalize the level of granularity we want between optimization functions, decompositions, etc (e.g. repeating ourselves vs function and class proliferation) (#80 ) 

We can put together a roadmap and collect thoughts/feedback on the project organization and API!

**Comment by caglayantuna** [comment]
HI @JeanKossaifi , thanks for the detailed roadmap. I think, it is necessary to have such a list for Tensorly developers. Slack works for me. Maybe, we could also inform each other about what we are working on to avoid time wasting. Thanks!

**Comment by cohenjer** [comment]
Great ! 
Slack works for me as well.
This roadmap looks excellent :) One decomposition that we handle kind of on the side right now is tensor trains and tensor networks in general, but supporting these properly would require a significant amount of work.
Also about having higher-end backend functions, one function that is kind of ``in the middle'' is the MTTKRP. Several research papers focus on how to compute it efficiently, but I don't think it is (and will be soon) implemented in the back-ends. A python implementation of it would be difficult since state-of-the-art MTTKRPs heavily use fine-tuned memory allocation and such (I am no expert ^^).

**Reply by JeanKossaifi** [reply]
I completely agree with both -- tensor networks should definitely be on the list and I've been adding more routines to manipulate TT/TTM. Allowing for various implementations is the goal of the tenalg backend that I hope to extend to allow for more optimized implementations.

**Comment by MarieRoald** [comment]
Thanks for such an excellent overview, @JeanKossaifi!  I agree with  @caglayantuna and @cohenjer that this list is very useful, and the https://github.com/tensorly/lab is a good idea! Slack works for me as well. 
I have a lot on my plate this summer, but I really look forward to working more on this when my schedule clears up :)

**Comment by aarmey** [comment]
This list is really useful @JeanKossaifi—there are many nice improvements here that I wasn't aware of. I really like the idea of a Slack. The Julia community one has been extremely helpful.

Would it be helpful to have the CMTF decomposition merged? I don't want to step on toes, but we're using this quite a bit, and having a tested reference implementation would be nice. I may be able to find someone in the lab to help with edits from the last review.

One longer-term thought... I know brute-optimization would add unwanted dependencies, but would it be helpful to think about adding common derivatives (e.g., the derivative of error between CP and a tensor)? This is another case of something we're already using and could allow a separate Tensorly package for brute optimization, or different line search routines, for example.

**Reply by IsabellLehmann** [reply]
Hi @aarmey , so do you have another implementation of CMTF? Fell free to create a pull request here - with my implementation there were always bugs with other than numpy backends.... Or do you mean if my implementation should be merged? The problem right now is that I have created some special functions for solving permutation and sign ambiguity to check if the decomposition was successful. If we just check the relative error (which I think is the case for the other decompositions), it might be easier to integrate,

**Reply by JeanKossaifi** [reply]
@IsabellLehmann, your PR was great and almost ready to merge - if you're able to reopen a PR perhaps we can build on that and finalize what needs to be done?

@aarmey I'm onboard for both points! Regarding the second part, do you mean writing the explicit gradient of strategic functions or having optimization routines? I definitely want to add the gradients, there are several cases where it's much better to hand compute them VS utilizing autograd. Regarding optimization routines I am very open to it, if cautious not to make the codebase too big to maintain/too general -- increase of complexity and inter-dependencies within the codebase also means that if anything breaks, so does the whole packages. The other thing is that often it is more efficient to tailor the optimization routine to the particular algorithm to make additional optimizations. So happy to give it a try, we could try in TensorLy-Lab or even by creating a separate repository for quick development?

**Reply by aarmey** [reply]
@IsabellLehmann I only have an implementation tailored to my with little to no testing. If you'd like I'm happy to help figure out the backend issues.

@JeanKossaifi great—I think I have some of the generally useful gradients coded up for CP. I'll think about appropriate tests and then make a draft PR.

**Reply by IsabellLehmann** [reply]
@aarmey I have created a pull request for the CMTF-ALS but I also do not have lots of testing... The code works right now only with the numpy backend because I used numpy.linalg.lstsq. If we have a tensorly version of least squares, it should work for all backends.


However, I realized that for a noisy tensor, the alternating least squares algorithm would probably not lead to a small absolute reconstruction error if we do not set the max_iter very high (maybe 50000) and the tolerance for the relative error change very small (maybe 1e-10?).

So in case you have a version of CMTF-OPT, it would be great to also have it in tensorly.

**Reply by cohenjer** [reply]
Actually we have recently worked with @MarieRoald and two other great researchers (Carla Schenker and Evrim Acar) on a ADMM framework for more flexible CMTF. Since it is based on the same AOADMM that @caglayantuna made a PR on recently, it would not be too hard to have it as well. We already have the proximal operators and a syntax for contraints is on its way, so sometime after summer when @MarieRoald has more time, we can work together with @caglayantuna to have AOADMM for coupled decompositions in Tensorly!

**Reply by aarmey** [reply]
@IsabellLehmann great! I should be able to add that to all backends. At least in my experience ALS has converged much faster than our implementation of CMTF-OPT. Your implementation looks pretty similar to [what I have](https://github.com/meyer-lab/systemsSerology/blob/e6d306761c18ed5c26b635ffb0d9bc6411b40bd0/syserol/tensor.py#L185). However, we do have an implementation of CMTF-OPT [here](https://github.com/meyer-lab/tfac-mrsa/blob/master/tfac/tensor.py) as the missing value pattern was showing up in the factors with ALS in one project.

@JeanKossaifi how would you advise we both access to a PR?

**Reply by yngvem** [reply]
> @JeanKossaifi how would you advise we both access to a PR?

I would add @IsabellLehmann's fork as a remote, fetch her CMTF_ALS branch and add a pull request into her CMTF_ALS branch. When it's accepted, it will automatically be added to the PR into tensorly main. @JeanKossaifi do you have any other approach?

**Reply by JeanKossaifi** [reply]
This sounds great @yngvem @aarmey, I think it may indeed be the easiest to first finish this PR so we're all on the same page as some of the features are reused and we'd want to have the same API for all CMTF versions. Do we need to explicitly allow others to add to the PR?

**Reply by aarmey** [reply]
I _think_ the three approaches are:
1. @JeanKossaifi can provide write permissions on Tensorly (optionally locking main and other branches for admins/maintainers only)
2. @IsabellLehmann can provide write permissions to her fork.
3. I can make a PR on @IsabellLehmann's fork.

I'll proceed with 3. It might help to think about the other options one day.

**Reply by IsabellLehmann** [reply]
Hi @aarmey. Thank you :) I have invited you as a collaborator of my fork.

**Comment by JeanKossaifi** [comment]
Hi all, I am slowly catching up with all the tasks!

I have created a Slack for the TensorLy project: [Join here](https://join.slack.com/t/newworkspace-bh44369/shared_invite/zt-u21984bp-O7jOb3pSOr4hzMLoDME0wA)

Looking forward to seeing you there! 

In the meantime, feel free to edit the top message or suggest anything else for the roadmap and happy summer!
